vendors/nuvoton/boards/numaker_iot_m487_wifi/aws_tests/application_code/main.c

```diff
@@ -0,0 +1,432 @@
+    configMAC_ADDR3,
+    configMAC_ADDR4,
+    configMAC_ADDR5
+};
+
+/**
+ * @brief Application task startup hook.
+ */
+void vApplicationDaemonTaskStartupHook( void );
+
+/**
+ * @brief Connects to Wi-Fi.
+ */
+static void prvWifiConnect( void );
+
+/**
+ * @brief Initializes the board.
+ */
+static void prvMiscInitialization( void );
+
+/*-----------------------------------------------------------*/
+/**
+ * @Set up the hardware ready to run this demo.
+ */
+static void prvSetupHardware( void );
+/*-----------------------------------------------------------*/
+
+
+/**
+ * @brief Application runtime entry point.
+ */
+#define mainCHECK_TASK_PRIORITY             ( tskIDLE_PRIORITY + 3UL )
+
+#define mainCHECK_TASK_STACK_SIZE           ( configMINIMAL_STACK_SIZE )
+#define mainCHECK_DELAY                     ( ( portTickType ) 5000 / portTICK_RATE_MS )
+
+
+int main( void )
+{
+    /* Perform any hardware initialization that does not require the RTOS to be
+     * running.  */
+    prvMiscInitialization();
+    configPRINTF( ( "FreeRTOS_IPInit\n") );	
+
+    /* Start the scheduler.  Initialization that requires the OS to be running,
+     * including the WiFi initialization, is performed in the RTOS daemon task
+     * startup hook. */ 
+         
+    /* A simple example to demonstrate key and certificate provisioning in
+     * microcontroller flash using PKCS#11 interface. This should be replaced
+     * by production ready key provisioning mechanism. */
+    vDevModeKeyProvisioning();     
+    configPRINTF( ( "vTaskStartScheduler\n" ) );  
+    vTaskStartScheduler();
+
+    return 0;
+}
+/*-----------------------------------------------------------*/
+
+static void prvSetupHardware( void )
+{
+    /* Unlock protected registers */
+    SYS_UnlockReg();
+
+    /* Set XT1_OUT(PF.2) and XT1_IN(PF.3) to input mode */
+    PF->MODE &= ~(GPIO_MODE_MODE2_Msk | GPIO_MODE_MODE3_Msk);
+
+    /* Enable External XTAL (4~24 MHz) */
+    CLK_EnableXtalRC(CLK_PWRCTL_HXTEN_Msk);
+
+    /* Waiting for 12MHz clock ready */
+    CLK_WaitClockReady( CLK_STATUS_HXTSTB_Msk);
+
+    /* Set core clock as PLL_CLOCK from PLL */
+    CLK_SetCoreClock(FREQ_192MHZ);
+
+    /* Set both PCLK0 and PCLK1 as HCLK/2 */
+    CLK->PCLKDIV = CLK_PCLKDIV_PCLK0DIV2 | CLK_PCLKDIV_PCLK1DIV2;
+
+    /* Enable IP clock */
+    CLK_EnableModuleClock(UART0_MODULE);
+    CLK_EnableModuleClock(EMAC_MODULE);
+
+    /* Select IP clock source */
+    CLK_SetModuleClock(UART0_MODULE, CLK_CLKSEL1_UART0SEL_HXT, CLK_CLKDIV0_UART0(1));
+
+    // Configure MDC clock rate to HCLK / (127 + 1) = 1.5 MHz if system is running at 192 MHz
+    CLK_SetModuleClock(EMAC_MODULE, 0, CLK_CLKDIV3_EMAC(127));
+
+    /* Update System Core Clock */
+    /* User can use SystemCoreClockUpdate() to calculate SystemCoreClock. */
+    SystemCoreClockUpdate();
+
+
+
+    /* Set GPB multi-function pins for UART0 RXD and TXD */
+    SYS->GPB_MFPH &= ~(SYS_GPB_MFPH_PB12MFP_Msk | SYS_GPB_MFPH_PB13MFP_Msk);
```

Comment:
Line 201 -- Line 216. It's very interesting to see memory mapped code got mixed up in main.c. There supposed to be driver functions that can be called in application code. Not a huge concern, but not very maintainable.

Reply:
@yuhui-zheng , it's code for multi-function pin & CLK init, it depends on what HW IPs & peripherals used in user application. Normally, we only wrap API for classic driver excluding CLK & pin. I will consider how to wrap this kind code.